File: chain/tests/retire.ts
/**
 * `retire` — the coin notices its own end.
 *
 * **THIS SUITE IS TERMINAL. It ends the ledger it runs on.** Retirement is
 * irreversible by design, so once these cases pass, `sync_m2` refuses on this
 * validator forever and the auction and sync suites have nothing to sync. That
 * is why `retire.ts` is excluded from the run-everything script in
 * `Anchor.toml` and lives behind `./stack test test-retire`. Run
 * **`./stack reset`** afterwards.
 *
 * If it finds an already-retired ledger it proves what it still can — that the
 * peg stays stopped and that a second `retire` is harmless — and skips the rest
 * rather than passing for the wrong reason.
 *
 * The silence window is `RETIREMENT_SILENCE_SECONDS` here rather than the
 * shipped year. Nothing about the mechanism cares how long it is; the shipped
 * value is long precisely because the flag cannot be taken back.
 */
import { expect } from 'chai'
import {
  Keypair,
  LAMPORTS_PER_SOL,
  Transaction,
  sendAndConfirmTransaction,
} from '@solana/web3.js'

import {
  BN,
  big,
  harness,
  bootstrap,
  mockFund,
  setMockM2,
  syncAccounts,
  assetMetas,
  placeBidAccounts,
  withdrawBidAccounts,
  RETIREMENT_SILENCE_SECONDS,
} from './enc-harness.js'

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms))

describe('retire', () => {
  const h = harness()

  const config = () => h.program.account.config.fetch(h.configPda)
  const printer = () => h.program.account.printer.fetch(h.printerPda)
  const retire = (caller = h.authority) =>
    h.program.methods
      .retire()
      .accounts({ config: h.configPda, printer: h.printerPda })
      .rpc()

  const sync = () =>
    h.program.methods.syncM2().accounts(syncAccounts(h)).remainingAccounts(assetMetas(h)).rpc()

  /** Push a fresh figure through, which restarts the silence clock. */
  async function syncNow() {
    const p = await printer()
    await setMockM2(h, big(p.m2Value) + 1_000_000n, p.m2ReleaseDate.toNumber() + 86_400)
    await sync()
  }

  /** Wait, by the **validator's** clock, until the program has been silent long enough. */
  async function waitOutTheSilence() {
    const until = Number((await printer()).lastSyncAt) + RETIREMENT_SILENCE_SECONDS
    for (let attempt = 0; attempt < 60; attempt++) {
      const now = await h.chainNow()
      if (now >= until) return
      await sleep(Math.min((until - now) * 1000 + 500, 5_000))
    }
    throw new Error('the validator clock never reached the end of the silence')
  }

  let alreadyRetired = false

  before(async function () {
    this.timeout(300_000)
    await bootstrap(h)
    alreadyRetired = (await config()).retired
  })

  it('starts its silence clock at genesis, not at zero', async function () {
    // Left at zero, the coin would be decades overdue for retirement the
    // instant it was born and the first stranger to look could end it.
    const p = await printer()
    const at = Number(p.lastSyncAt)
    expect(at, 'the retirement clock was never started').to.be.greaterThan(1_600_000_000)
    expect(at, 'the retirement clock is in the future').to.be.at.most((await h.chainNow()) + 5)
  })

  it('refuses while the program has heard about money recently', async function () {
    if (alreadyRetired) this.skip()
    this.timeout(120_000)
    await syncNow()
    const why = await h.failureOf(() => retire())
    expect(why, 'the coin was retired while its peg was current').to.match(/NotSilentEnough/)
  })

  it('needs no signer beyond whoever pays the fee', async function () {
    if (alreadyRetired) this.skip()
    // The account list is two PDAs and nothing else — no authority, no owner,
    // no key of ours. Asserted against the shape of the instruction rather
    // than its behaviour, because "there is no key" is a claim about the shape.
    const ix = h.program.idl.instructions.find((i) => i.name === 'retire')
    expect(ix, 'retire is missing from the interface').to.not.be.undefined
    const signers = (ix!.accounts as { name: string; signer?: boolean }[]).filter((a) => a.signer)
    expect(signers.map((a) => a.name), 'retire consults a signer').to.be.empty
  })

  it('lets a stranger end it once the silence has run, and stops the peg', async function () {
    if (alreadyRetired) this.skip()
    this.timeout(180_000)
    await waitOutTheSilence()

    // A wallet with no relationship to this program at all: no bootstrap
    // authority, no tenancy, nothing but a fee.
    //
    // Sent as a raw transaction with the stranger as **fee payer**, rather than
    // through `.signers([...])`. The instruction takes no signer account at
    // all, so Anchor rejects an extra one outright ("unknown signer") — which
    // is itself the proof, and this is the only way to actually execute it as
    // somebody else.
    const stranger = Keypair.generate()
    const drop = await h.connection.requestAirdrop(stranger.publicKey, LAMPORTS_PER_SOL)
    await h.history.confirmTransaction(drop, 'confirmed')

    const before = await printer()
    const ix = await h.program.methods
      .retire()
      .accounts({ config: h.configPda, printer: h.printerPda })
      .instruction()
    await sendAndConfirmTransaction(h.connection, new Transaction().add(ix), [stranger], {
      commitment: 'confirmed',
    })

    const after = await config()
    expect(after.retired, 'the coin did not retire').to.equal(true)
    // The final numbers are left exactly as the Fed last reported them. The
    // auctions go on trading at these forever.
    const still = await printer()
    expect(still.m2Value.toString()).to.equal(before.m2Value.toString())
    expect(still.m2ReleaseDate.toString()).to.equal(before.m2ReleaseDate.toString())
  })

  it('refuses every later sync, forever', async function () {
    this.timeout(120_000)
    expect((await config()).retired, 'this case needs a retired ledger').to.equal(true)
    const p = await printer()
    // A perfectly good, genuinely newer release — refused because the peg has
    // stopped, not because anything is wrong with the number.
    await setMockM2(h, big(p.m2Value) + 1_000_000n, p.m2ReleaseDate.toNumber() + 86_400)
    const why = await h.failureOf(sync)
    expect(why, 'the peg kept running after retirement').to.match(/Retired/)
  })

  it('is harmless to call twice', async () => {
    expect((await config()).retired).to.equal(true)
    // No error: "anyone can walk up and observe that it ended" should not throw
    // at the second person who looks.
    await retire()
    expect((await config()).retired).to.equal(true)
  })

  it('goes on auctioning flags at the prices of a vanished world', async function () {
    this.timeout(180_000)
    // The ending Ruling B leaned toward, and it costs nothing: `PriceCurve`
    // already flattens at `price_to`, so after the final sync the machine holds
    // those numbers and the auction keeps running on them.
    //
    // This is also the *safety* case, not just the poetry. A freeze ending
    // would have had to keep `withdraw_bid` and every other escrow exit alive
    // anyway, or it would strand live bids permanently — the exact harm the
    // auction's criteria forbid. So the thing to prove is not that the
    // instructions still exist in the interface; it is that **money still gets
    // out**. Proven by doing it, on a retired chain.
    expect((await config()).retired, 'this case needs a retired ledger').to.equal(true)

    const i = 0
    const alice = Keypair.generate()
    const bob = Keypair.generate()
    for (const who of [alice, bob]) {
      const drop = await h.connection.requestAirdrop(who.publicKey, 2 * LAMPORTS_PER_SOL)
      await h.history.confirmTransaction(drop, 'confirmed')
    }

    // Every term has long since lapsed while we sat out the silence. Rolling it
    // is the same permissionless path a stranger would use, and it still works.
    const lapsed = await h.program.account.asset.fetch(h.assetPda(i))
    if ((await h.chainNow()) >= Number(lapsed.termEndsAt)) {
      await h.program.methods.rollTerm(i).accounts({ asset: h.assetPda(i) }).rpc()
    }

    const asset = await h.program.account.asset.fetch(h.assetPda(i))
    const reserve = big(asset.priceTo)
    expect(reserve > 0n, 'the last price the Fed ever reported was zero').to.equal(true)
    const stake = reserve * 4n
    for (const who of [alice, bob]) await mockFund(h, who.publicKey, stake)

    // Both bids clear a reserve nothing will ever move again.
    const first = reserve + reserve / 1_000n
    await h.program.methods
      .placeBid(i, new BN(first.toString()))
      .accounts(placeBidAccounts(h, i, alice.publicKey))
      .signers([alice])
      .rpc()
    await h.program.methods
      .placeBid(i, new BN((first + 1_000n).toString()))
      .accounts(placeBidAccounts(h, i, bob.publicKey))
      .signers([bob])
      .rpc()

    // And the superseded bidder gets her money back, after the end of the
    // world. No escrow is stranded by retirement.
    const balance = await h.connection.getTokenAccountBalance(h.encAta(alice.publicKey))
    await h.program.methods
      .withdrawBid(i)
      .accounts(withdrawBidAccounts(h, i, alice.publicKey))
      .signers([alice])
      .rpc()
    const after = await h.connection.getTokenAccountBalance(h.encAta(alice.publicKey))
    expect(
      BigInt(after.value.amount) - BigInt(balance.value.amount),
      'escrow was stranded by retirement',
    ).to.equal(first)

    // And the faucet is untouched too — retirement stops the peg, not the game.
    expect((await config()).retired).to.equal(true)
  })
})

File: chain/programs/emperors-new-coin/src/instructions/retire.rs
//! The coin notices its own end.
//!
//! ENC runs **forever** (Ruling B, 2026-08-12) — with one exception it can
//! prove for itself. If a full `retirement_silence_seconds` passes in which no
//! new M2 figure reaches this program, anyone may call `retire`: no key is
//! consulted, no announcement is made, and a passer-by can simply observe that
//! it is over. Nobody can end it early; nobody can prevent it once the
//! condition is true.
//!
//! **Why elapsed time, and not "the value stopped changing".** The program has
//! no clock of its own and only runs when somebody sends it a transaction. A
//! staleness *counter* would need someone to keep poking it. An elapsed-time
//! test is true whether or not anyone is watching, so the first visitor years
//! later can flip the bit.
//!
//! **What a year of silence actually means — and the copy must not overclaim.**
//! The trigger measures exactly one thing: *no new M2 figure reached this
//! program for a year.* The program cannot know why. Over a forever horizon the
//! likely causes, roughly in order: the oracle stack rotted (Switchboard
//! sunsets the feed's runtime, against a pinned feed hash on a non-upgradeable
//! program — near-certain eventually); nobody left who cared to crank; FRED
//! stopped serving; the Fed went dark. Oracle silence is **not** Fed silence.
//! What is true is also the better joke: **the coin ends when nobody has told
//! it about money for a year — whether because the dollar ended or because
//! everyone stopped looking, and from where it sits those are the same event.**
//!
//! **What retirement does, and does not, stop.** It stops `sync_m2`, and that
//! is all. The auctions keep running forever at the last prices the Fed ever
//! reported — the machine grinding on, trading flags at the valuations of a
//! vanished world, because nobody noticed the numbers stopped meaning anything.
//! That is free: `PriceCurve` already flattens at `price_to`, so after the
//! final sync the machine simply holds. It is also the *safe* ending — a freeze
//! would have had to keep `withdraw_bid` and every other escrow exit alive
//! anyway, or it would strand live bids permanently, which is the exact harm
//! the auction's criteria forbid.
use anchor_lang::prelude::*;

use crate::errors::EncError;
use crate::state::*;

/// Emitted once, ever. The receipt for the end of the artwork.
#[event]
pub struct Retired {
    /// When the bit was flipped.
    pub at: i64,
    /// The last time anyone told this program what money was.
    pub last_sync_at: i64,
    /// What the Fed last said, and when they said it — the numbers the
    /// auctions go on trading at forever.
    pub final_m2_value: u64,
    pub final_m2_release_date: i64,
}

pub fn handler(ctx: Context<Retire>) -> Result<()> {
    let now = Clock::get()?.unix_timestamp;
    let config = &mut ctx.accounts.config;
    let printer = &ctx.accounts.printer;

    let silent_since = now
        .checked_sub(printer.last_sync_at)
        .ok_or(error!(EncError::MathOverflow))?;
    require!(
        silent_since >= config.retirement_silence_seconds,
        EncError::NotSilentEnough
    );

    // Deliberately not guarded against a repeat. Calling it twice is harmless —
    // the condition is still true, the bit is already set, and refusing would
    // make "anyone can walk up and observe that it ended" throw an error at the
    // second person who looks.
    config.retired = true;

    emit!(Retired {
        at: now,
        last_sync_at: printer.last_sync_at,
        final_m2_value: printer.m2_value,
        final_m2_release_date: printer.m2_release_date,
    });

    msg!(
        "retired after {silent_since}s of silence; the last word on money was {} on {}",
        printer.m2_value,
        printer.m2_release_date
    );
    Ok(())
}

/// No signer at all, and no authority anywhere. The fee payer signs the
/// transaction; this instruction asks nothing of anyone, which is the whole
/// point of an ending nobody has to be trusted to declare.
#[derive(Accounts)]
pub struct Retire<'info> {
    #[account(mut, seeds = [CONFIG_SEED], bump = config.bump)]
    pub config: Account<'info, Config>,

    #[account(seeds = [PRINTER_SEED], bump = printer.bump)]
    pub printer: Account<'info, Printer>,
}
